pkg/security/client.go

```diff
@@ -156,34 +160,126 @@ func (c *Client) getSecurityPolicy(ctx context.Context, cluster *capg.GCPCluster
 	return c.securityPolicies.Get(ctx, req)
 }
 
-func (c *Client) updateSecurityPolicy(ctx context.Context, cluster *capg.GCPCluster, policy *computepb.SecurityPolicy) error {
+func (c *Client) updateSecurityPolicy(ctx context.Context, cluster *capg.GCPCluster, policy *computepb.SecurityPolicy) (*computepb.SecurityPolicy, error) {
+	currentPolicy, err := c.getSecurityPolicy(ctx, cluster, *policy.Name)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	// There are three groups of rules - new rules, rules we want to update and
+	// rules that we want to delete. Rules that are in the current policy, but
+	// not in the new one should be deleted.
+	// We start of by marking all the rules in the current policy for deletion.
+	rulesToDelete := constructRulePriorityMap(currentPolicy.Rules)
 	for _, rule := range policy.Rules {
-		req := &computepb.PatchRuleSecurityPolicyRequest{
-			Priority:                   rule.Priority,
-			Project:                    cluster.Spec.Project,
-			SecurityPolicy:             *policy.Name,
-			SecurityPolicyRuleResource: rule,
+		priority := *rule.Priority
+
+		// If both the new and old policy contain a rule with the same
+		// priority, then patch that rule and remove it from the rules that
+		// need to be deleted.
+		// If a rule is only in the new policy then it needs to be created.
+		// NOTE: We can't check for 404 when patching the rule. GCP will always
+		// return 400 if a rule doesn't exist. This also applies to `GetRule`
+		_, ok := rulesToDelete[priority]
+		if ok {
+			delete(rulesToDelete, priority)
+			err = c.patchRule(ctx, cluster, policy, rule)
+			if err != nil {
+				return nil, errors.WithStack(err)
+			}
+
+			continue
 		}
-		op, err := c.securityPolicies.PatchRule(ctx, req)
+
+		err = c.createRule(ctx, cluster, policy, rule)
 		if err != nil {
-			return errors.WithStack(err)
+			return nil, errors.WithStack(err)
+		}
+	}
+
+	for rulePriority := range rulesToDelete {
+		// The default has priority MaxInt32 and cannot be deleted
+		if rulePriority == math.MaxInt32 {
```

Comment:
Yeah I think that's way better. 

How do you feel about the rest of the comments. I thought they're necessary to help anyone looking at it, because it's not 100% clear what's happening and why, but I'm not really sure.

Reply:
The other comments helped me understand what was going on. Even if we come up with better function names, I don't know if that would be as good as the comments, for that particular case.

Reply:
Alright, I've added a constant for the default priority. I think it gets the point across now

Reply:
Apparently `math.MaxInt32` is an `int` instead of `int32` :unamused:

Reply:
_Obviously_